File: shift/tx.go
package main

/*
 * tx.go
 * Goroutine to send data to insert
 * by J. Stuart McMurray
 * created 20150116
 * last modified 20150126
 *
 * Copyright (c) 2014 J. Stuart McMurray <kd5pbo@gmail.com>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

import (
	"crypto/rand"
	"encoding/binary"
	"fmt"
	"log"
	"math"
	"math/big"
	"time"
)

var (
	kaMin = 2    /* Minimum keepalive junk length */
	kaMax = 1024 /* Maximum  keepalive junk length */
)

/* Errors that might be sent on echan */
var (
	ErrorKATooBig = fmt.Errorf("Keepalive size is larger than a 16-bit uint")
)

/* Reads from the tun device and sends the data to insert. fatal errors will be
reported on echan.  The goroutine will terminate when dchan is closed. */
func tx(
	tun Tunnel,
	in *Insert,
	echan chan error,
	/* Time range to wait before sending a keepalive */
	minWait time.Duration,
	maxWait time.Duration,
) {
	/* Start reads from the tunnel into a chan */
	fchan := make(chan Frame)
	go readIntoChan(tun, fchan, echan)

	for {
		/* Bounded random wait before sending keepalive */
		wait, err := randomWait(minWait, maxWait)
		if nil != err {
			echan <- err
			return
		}
		/* Wait for either time to send a keepalive or a packet */
		select {
		case <-time.After(wait): /* Send a keepalive */
			if err := sendKeepalive(in); nil != err {
				echan <- err
				return
			}

		case f, ok := <-fchan: /* (Maybe) send a frame */
			/* Give up if the channel's closed */
			if !ok {
				return
			}
			/* Try to send the frame to insert */
			if err := sendToInsert(
				in,
				f,
				tun.MaxFrameLen(),
			); nil != err {
				echan <- err
				return
			}
		}
	}
}

/* Read from a Tunnel into a chan, which will be closed on error (which will
be sent to echan) */
func readIntoChan(t Tunnel, fchan chan Frame, echan chan error) {
	for {
		/* Read a frame from the tunnel */
		f, err := t.Read()
		/* Give up if there's an error */
		if nil != err {
			echan <- err
			close(fchan)
			return
		}
		/* Send the frame on the channel */
		fchan <- f
	}
}

/* Marshall a frame and send it to the insert if it's shorter than maxLen  */
func sendToInsert(in *Insert, f Frame, maxLen int) error {
	/* Drop frames that are bigger than the tunnel can handle */
	if maxLen < len(f) {
		log.Printf(
			"Dropping frame of length %v > %v",
			len(f),
			maxLen,
		)
		return nil
	}
	/* Drop frames that are bigger than the protocol can handle */
	if math.MaxUint16 < len(f) {
		log.Printf(
			"Dropping %v-byte frame that is bigger than "+
				"the protocol's max %v bytes.",
			len(f),
			math.MaxUint16,
		)
		return nil
	}
	/* Marshall a nice message */
	mf, err := f.Marshall()
	if nil != err { /* Shouldn't happen */
		log.Printf("Marshall error: %v", err)
		return nil
	}
	/* Send frame to insert */
	if err := in.SendEnc(mf); nil != err {
		return err
	}

	return nil
}

/* randomWait returns a time.Duration between min and max */
func randomWait(min, max time.Duration) (time.Duration, error) {
	big, err := rand.Int(rand.Reader, big.NewInt(int64(max-min)))
	if nil != err {
		return 0, err
	}
	return time.Duration(big.Int64()) + min, nil
}

/* sendKeepalive sends a keepalive to Insert */
func sendKeepalive(in *Insert) error {
	/* Get a random size */
	big, err := rand.Int(rand.Reader, big.NewInt(int64(kaMax-kaMin)))
	if nil != err {
		return err
	}
	sizeh := big.Uint64() + uint64(kaMin)

	/* Make sure it's not too large */
	if math.MaxUint16 < sizeh {
		return ErrorKATooBig
	}

	/* Get it network byte order */
	sizen := make([]byte, 2)
	binary.BigEndian.PutUint16(sizen, uint16(sizeh))

	/* Make that much junk */
	junk := make([]byte, sizeh)
	if _, err := rand.Read(junk); nil != err {
		return err
	}

	/* Put it all together and send it */
	header := append([]byte{0x00, 0x00}, sizen...)
	ka := append(header, junk...)
	debug("Sending %v-byte keepalive", sizeh)
	if err := in.SendEnc(ka); nil != err {
		return err
	}

	return nil
}

File: insert/tx.c
/*
 * tx.c
 * Code for thread to send data to insert
 * by J. Stuart McMurray
 * created 20150212
 * last modified 20150222
 *
 * Copyright (c) 2015 J. Stuart McMurray <kd5pbo@gmail.com>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include <netinet/in.h>
#include <string.h>

#include "comm.h"
#include "insert.h"
#include "retvals.h"
#include "sha2.h"
#include "tx.h"

/* Get data from pcap, send to shift */
void *insert_to_shift(void *data) {
        struct its_data id;    /* Input data, pulled from the void* */
        struct handle_data hd; /* Data for handle_frame */
        int ret;               /* Return value */
        int hret;              /* Return value from handler */

        memset(&hd, 0, sizeof(hd));
        ret = 0;
        hret = 0;

        /* Get a copy of the input data */
        memcpy(&id, data, sizeof(id));

        /* Data for handle_frame */
        hd.p = id.p;
        hd.fd = id.fd;
        hd.ret = &hret;

        /* Capture frames until an error occurs */
        ret = pcap_loop(id.p, 0, handle_packet, (u_char*)&hd);
        switch (ret) {
                case -2: /* Explicit loop break */
                        set_reterr(hret);
                        return NULL;
                        break;
                case -1:
                        set_reterr(RET_ERR_CAP);
                        return NULL;
                        break;
                default:
                        printf("Unknown pcap_loop return: %i\n", ret);
                        set_reterr(RET_ERR_CAP);
                        return NULL;
                        break;
        }

        /* Shouldn't get here */
        return NULL;
}

/* Callback function for pcap_loop */
void handle_packet(u_char *user, const struct pcap_pkthdr *header,
        const u_char *data) {

        uint8_t databuf[UINT16_MAX+sizeof(uint16_t)];
                                   /* Buffer to hold the data */
        uint8_t txhash[DIGESTLEN]; /* Hash of the data */
        uint16_t len;              /* Size of the data */
        struct handle_data hd;     /* User data passed in */
        int ret;                   /* Return value */

        memset(txhash, 0, sizeof(txhash));
        len = 0;
        memcpy(&hd, user, sizeof(hd));

        /* Make sure we captured the entire frame */
        if (header->len != header->caplen) {
                *hd.ret = RET_ERR_CSZS;
                goto BREAK;
        }

        /* Make sure frame isn't too large */
        if (UINT16_MAX < header->len) {
                *hd.ret = RET_ERR_CSZL;  
                goto BREAK;
        }

        /* Copy the size to a buffer */
        len = htons(header->len);
        memcpy(databuf, &len, sizeof(len));

        /* Copy the data to a buffer */
        memcpy(databuf+sizeof(len), data, header->len);

        /* Calculate hash */
        sha224(databuf, header->len+sizeof(len), txhash);



        /* Send the bits.  We're boned if anything else is sending */
        if ((0 != (ret = send_enc(hd.fd, databuf, header->len+sizeof(len)))) ||
                        (0 != (ret = send_enc(hd.fd, txhash,
                                              sizeof(txhash))))) {
                *hd.ret = ret;
                goto BREAK;
        }

        /* Return success */
        *hd.ret = 0;
        return;

        /* Something bad happened, stop the capture */
BREAK:
        pcap_breakloop(hd.p);
        return;
}
